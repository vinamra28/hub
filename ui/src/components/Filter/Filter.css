.guruji {
  margin: 3em;
}
.foo {
  width: 10em;
}
.foobar {
  padding-left: 0.7em;
  margin-top: 0.5em;
}
