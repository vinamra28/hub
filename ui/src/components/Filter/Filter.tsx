import React from 'react';
import { Checkbox } from '@patternfly/react-core';
import { useObserver } from 'mobx-react';
import { Text, TextVariants } from '@patternfly/react-core';
import { Flex, FlexItem } from '@patternfly/react-core';
import { Button } from '@patternfly/react-core/dist/js/components';
import TimesIcon from '@patternfly/react-icons/dist/js/icons/times-icon';
import '@patternfly/react-core/dist/styles/base.css';

interface Filterable {
  id: number;
  name: string;
  selected: boolean;
  toggle(): void;
}

interface Store {
  items: Filterable[];
  clear(): void;
}

interface FilterList {
  store: Store;
  header: string;
}

const checkboxes = (items: Filterable[]) =>
  items.map((c: Filterable) => (
    <Checkbox
      key={c.id}
      label={c.name}
      isChecked={c.selected}
      onChange={() => c.toggle()}
      aria-label="controlled checkbox"
      id={`${c.id}`}
      name={c.name}
    />
  ));

const Filter: React.FC<FilterList> = ({ store, header }) => {
  return useObserver(() => (
    <div style={{ margin: '5em' }}>
      <Flex>
        <FlexItem>
          <Text component={TextVariants.h1} style={{ fontWeight: 'bold' }}>
            {header}
          </Text>
        </FlexItem>
        <FlexItem>
          <Button variant="plain" aria-label="Clear" onClick={store.clear}>
            <TimesIcon />
          </Button>
          <br />
        </FlexItem>
      </Flex>

      <Flex>
        <FlexItem>
          <div>{checkboxes(store.items)}</div>
        </FlexItem>
      </Flex>
    </div>
  ));
};

export default Filter;
