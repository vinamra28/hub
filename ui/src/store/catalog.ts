import { types, Instance } from 'mobx-state-tree';

export const CatalogType = types
  .model({
    id: types.integer,
    name: types.string,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICatalogType = Instance<typeof CatalogType>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model('CatalogStore', {
    list: types.array(CatalogType)
  })

  .views((self) => ({
    get count(): number {
      return self.list.length;
    },
    get catalogs(): Array<string> {
      const { list } = self;
      return list
        .filter((type: ICatalogType) => type.selected)
        .map((type: ICatalogType) => type.name);
    }
  }))

  .actions((self) => ({
    add(item: ICatalogType) {
      self.list.push(item);
    },
    clear() {
      self.list.map((item: ICatalogType) => (item.selected = false));
    }
  }));
