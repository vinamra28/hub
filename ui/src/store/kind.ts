import { types, Instance } from 'mobx-state-tree';

export const Kind = types
  .model('Kind', {
    id: types.integer,
    name: types.string,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type IKind = Instance<typeof Kind>;
export type IKindStore = Instance<typeof KindStore>;

export const KindStore = types
  .model('KindStore', {
    list: types.array(Kind)
  })

  .views((self) => ({
    get count(): number {
      return self.list.length;
    },
    get kinds(): Array<string> {
      const { list } = self;
      return list.filter((type: IKind) => type.selected).map((type: IKind) => type.name);
    }
  }))

  .actions((self) => ({
    add(item: IKind) {
      self.list.push(item);
    },
    clear() {
      self.list.map((item: IKind) => (item.selected = false));
    }
  }));
