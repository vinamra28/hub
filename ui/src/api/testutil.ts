import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource } from '../store/resources';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
