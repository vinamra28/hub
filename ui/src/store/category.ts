import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Tag = types.model('Tags', {
  id: types.integer,
  name: types.string
});

export const Category = types
  .model('Category', {
    id: types.number,
    name: types.string,
    tags: types.array(Tag),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICategory = Instance<typeof Category>;
export type ITag = Instance<typeof Tag>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.array(Category),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.length;
    },

    get tags() {
      return self.items
        .filter((c) => c.selected)
        .reduce((acc: string[], c: ICategory) => [...acc, ...c.tags.map((t) => t.name)], []);
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.push(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clear() {
      self.items.map((c: ICategory) => (c.selected = false));
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;
        const json = yield api.categories();
        json.data.forEach((item: ICategory) => self.add(item));
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));

export type ICategoryStore = Instance<typeof CategoryStore>;
