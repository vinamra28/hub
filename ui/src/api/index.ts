import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource } from '../store/resource';

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: string) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
