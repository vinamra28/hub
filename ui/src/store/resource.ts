import { types, Instance } from 'mobx-state-tree';
import fuzzysort from 'fuzzysort';
import moment, { Moment } from 'moment';
import { flow, getEnv } from 'mobx-state-tree';
import { Tag, ICategoryStore, ITag } from './category';
import { Api } from '../api';
import { Catalog, CatalogStore } from './catalog';
import { Kind, KindStore } from './kind';

export const updatedAt = types.custom<string, Moment>({
  name: 'momentDate',
  fromSnapshot(value: string): Moment {
    return moment(new Date(value));
  },
  toSnapshot(value: Moment): string {
    return value.fromNow();
  },
  isTargetType(v: string) {
    return moment.isMoment(v);
  },
  getValidationMessage(v: string) {
    if (moment.isMoment(v)) {
      return 'Invalid moment object';
    }
    return '';
  }
});

const Version = types.model('Version', {
  id: types.identifierNumber,
  version: types.string,
  displayName: types.string,
  description: types.string,
  minPipelinesVersion: types.string,
  rawURL: types.string,
  webURL: types.string,
  updatedAt: updatedAt
});

export const Resource = types.model('Resource', {
  id: types.number,
  name: types.identifier,
  catalog: types.reference(Catalog),
  kind: types.reference(Kind),
  latestVersion: types.reference(Version),
  tags: types.array(types.reference(Tag)), // ["1", "2"]
  rating: types.number,
  versions: types.array(types.reference(Version)),
  displayName: ''
});

export type IResource = Instance<typeof Resource>;
export type IVersion = Instance<typeof Version>;

export enum sortByFields {
  Name = 'Name',
  Rating = 'Rating'
}

export const ResourceStore = types
  .model('ResourceStore', {
    resources: types.map(Resource),
    versions: types.map(Version),
    catalogs: types.optional(CatalogStore, {}),
    kinds: types.optional(KindStore, {}),
    sortBy: types.optional(types.enumeration(Object.values(sortByFields)), sortByFields.Name),
    tags: types.optional(types.map(Tag), {}),
    search: '',
    err: '',
    isLoading: true
  })
  .views((self) => ({
    get items() {
      return Array.from(self.resources.values());
    },
    get api(): Api {
      return getEnv(self).api;
    },
    get categories(): ICategoryStore {
      return getEnv(self).categories;
    }
  }))

  .actions((self) => ({
    setLoading(l: boolean) {
      self.isLoading = l;
    },
    setSearch(text: string) {
      self.search = text;
    },
    add(item: IResource) {
      self.resources.put(item);
    },
    setSortBy(field: string) {
      const key: sortByFields = sortByFields[field as keyof typeof sortByFields];
      self.sortBy = key;
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);

        const { api } = self;
        const json = yield api.resources();

        const kinds: string[] = json.data.map((r: IResource) => r.kind);
        kinds.forEach((k) => self.kinds.add(k));

        json.data.forEach((r: IResource) => {
          self.catalogs.add(r.catalog);
          self.versions.put(r.latestVersion);
        });

        // adding the tags to the store - normalized
        const tags: ITag[] = json.data.flatMap((item: IResource) => item.tags);

        tags.forEach((t) => (t != null ? self.tags.put(t) : null));

        const resources: IResource[] = json.data.map((r: IResource) => ({
          id: r.id,
          name: r.name,
          catalog: r.catalog.id,
          kind: r.kind,
          latestVersion: r.latestVersion.id,
          tags: r.tags != null ? r.tags.map((tag: ITag) => tag.id) : [],
          rating: r.rating,
          versions: [],
          displayName: r.latestVersion.displayName
        }));

        resources.forEach((r: IResource) => {
          r.versions.push(r.latestVersion);
          self.add(r);
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }))

  .views((self) => ({
    get filteredResources() {
      const { resources, kinds, catalogs, categories, search } = self;
      const { selectedTags } = categories;

      const filtered: IResource[] = [];
      resources.forEach((r: IResource) => {
        const matchesKind = kinds.selected.size === 0 || kinds.selected.has(r.kind.name);
        const matchesCatalogs = catalogs.selected.size === 0 || catalogs.selected.has(r.catalog.id);
        const matchesTags = selectedTags.size === 0 || r.tags.some((t) => selectedTags.has(t.id));

        if (matchesKind && matchesCatalogs && matchesTags) {
          filtered.push(r);
        }
      });

      if (search.trim() !== '') {
        return fuzzysort
          .go(search, filtered, { keys: ['name', 'displayName'] })
          .map((resource: Fuzzysort.KeysResult<IResource>) => resource.obj);
      }
      return filtered;
    }
  }));

export type IResourceStore = Instance<typeof ResourceStore>;
