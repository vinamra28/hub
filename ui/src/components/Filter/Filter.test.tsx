import React from 'react';
import { FakeHub } from '../../api/testutil';
import { when } from 'mobx';
import { shallow } from 'enzyme';
import Filter from './Filter';
import { CategoryStore } from '../../store/category';

const TESTDATA_DIR = `src/store/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Filter component', () => {
  it('should render the component', (done) => {
    const store = CategoryStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        const component = shallow(<Filter store={store} header="Categories" />);
        expect(component).toMatchSnapshot();

        done();
      }
    );
  });

  it('checks the checkbox render count', (done) => {
    const store = CategoryStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        const component = shallow(<Filter store={store} header="Categories" />);
        expect(component.find('Checkbox[aria-label="controlled checkbox"]').length).toEqual(5);

        done();
      }
    );
  });

  it('finds the button and matches the count', (done) => {
    const store = CategoryStore.create({}, { api });

    when(
      () => !store.isLoading,
      () => {
        const component = shallow(<Filter store={store} header="Categories" />);

        expect(component.find('Button[variant="plain"]').length).toEqual(1);

        done();
      }
    );
  });
});
