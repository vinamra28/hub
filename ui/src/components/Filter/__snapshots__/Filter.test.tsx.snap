// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component should render the component 1`] = `ShallowWrapper {}`;
